be/src/olap/rowset/segment_v2/segment_iterator.cpp

```diff
@@ -2138,22 +2143,21 @@ void SegmentIterator::_build_index_result_column(uint16_t* sel_rowid_idx, uint16
     size_t idx_in_row_range = 0;
     size_t idx_in_selected = 0;
     // _split_row_ranges store multiple ranges which split in function _read_columns_by_index(),
-    // index_result is a column predicate apply result in a whole segement,
-    // but a scanner thread one time can read max rows limit by block_row_max,
+    // index_result is a column predicate apply result in a whole segment,
+    // but a scanner thread can read max rows limit by block_row_max one time,
     // so split _row_bitmap by one time scan range, in order to match size of one scanner thread read rows.
-    for (auto origin_row_range : _split_row_ranges) {
-        for (size_t rowid = origin_row_range.first; rowid < origin_row_range.second; ++rowid) {
-            if (sel_rowid_idx == nullptr || (idx_in_selected < select_size &&
-                                             idx_in_row_range == sel_rowid_idx[idx_in_selected])) {
-                if (index_result.contains(rowid)) {
-                    vec_match_pred[idx_in_block++] = true;
-                } else {
-                    vec_match_pred[idx_in_block++] = false;
-                }
-                idx_in_selected++;
+    for (uint32_t idx = 0; idx < _current_batch_rows_read; idx++) {
+        auto rowid = _block_rowids[idx];
+        if (sel_rowid_idx == nullptr ||
+            (idx_in_selected < select_size && idx_in_row_range == sel_rowid_idx[idx_in_selected])) {
```

Comment:
The original logic `idx_in_row_range == sel_rowid_idx[idx_in_selected])` indicate that data in sel_rowid_idx is relative index in read block. It need to be verified once more.